### PR TITLE
Fixed #25759 -- Added keyword arguments to customize Expressions' as_sql().

### PR DESCRIPTION
**Pull request by timgraham**
Rebased from https://github.com/django/django/pull/6302.

### CODE REVIEW

File: django/db/models/expressions.py [hunk 1]
```diff
@@ -542,16 +542,22 @@ def as_sql(self, compiler, connection, function=None, template=None):
...
-
-    def as_sqlite(self, *args, **kwargs):
-        sql, params = self.as_sql(*args, **kwargs)
+            data.setdefault('function', self.function)
+        template = template or data.get('template', self.template)
+        arg_joiner = arg_joiner or data.get('arg_joiner', self.arg_joiner)
+        data['expressions'] = data['field'] = arg_joiner.join(sql_parts)
+        return template % data, params
+
+    def as_sqlite(self, compiler, connection):
```
Comment (timgraham) [inline django/db/models/expressions.py]: I'm thinking this signature change should go in a separate commit (assuming  it's okay).

Comment (charettes) [inline django/db/models/expressions.py]: Makes sense to me.

File: tests/db_functions/models.py [hunk 2]
```diff
@@ -41,3 +41,12 @@ class Fan(models.Model):
 
     def __str__(self):
         return self.name
+
+
+@python_2_unicode_compatible
+class Number(models.Model):
+    integer = models.BigIntegerField()
+    float = models.FloatField(null=True)
```
Comment (charettes) [inline tests/db_functions/models.py]: Apart from being used in `__str__` and `setUpTestData` the `float` field seems unused. We should either add an extra test to cast it somehow or simply remove it.

**Comment by charettes** [comment]
Buildbot, test on oracle.

File: tests/db_functions/models.py (hunk 2 above)
Comment (timgraham) [inline tests/db_functions/models.py]: Thanks, I reused an existing model.
